apps/studio/components/layouts/ProjectLayout/LayoutHeader/NotificationsPopoverV2/NotificationsPopover.tsx

```diff
@@ -0,0 +1,265 @@
+  IconAlertCircle,
+  IconAlertTriangle,
+  IconCheck,
+  IconInbox,
+  PopoverContent_Shadcn_,
+  PopoverTrigger_Shadcn_,
+  Popover_Shadcn_,
+  Tabs,
+} from 'ui'
+
+import AlertError from 'components/ui/AlertError'
+import InfiniteList from 'components/ui/InfiniteList'
+import ShimmeringLoader, { GenericSkeletonLoader } from 'components/ui/ShimmeringLoader'
+import { useNotificationsV2Query } from 'data/notifications/notifications-v2-query'
+import { useNotificationsV2UpdateMutation } from 'data/notifications/notifications-v2-update-mutation'
+import { useOrganizationsQuery } from 'data/organizations/organizations-query'
+import { useProjectsQuery } from 'data/projects/projects-query'
+import NotificationRow from './NotificationRow'
+import { NOTIFICATION_FILTERS, NOTIFICATION_FILTER_TYPE } from './NotificationsPopover.constants'
+import { useNotificationsSummaryQuery } from 'data/notifications/notifications-v2-summary-query'
+
+const NotificationsPopverV2 = () => {
+  const [open, setOpen] = useState(false)
+  const [openFilters, setOpenFilters] = useState(false)
+  const [selectedFilter, setSelectedFilter] = useState<NOTIFICATION_FILTER_TYPE>('all')
+  const [activeTab, setActiveTab] = useState<'inbox' | 'archived'>('inbox')
+
+  // [Joshen] Just FYI this variable row heights logic should ideally live in InfiniteList
+  // but I ran into some infinite loops issues when I was trying to implement it there
+  // so opting to simplify and implement it here for now
+  const rowHeights = useRef<{ [key: number]: number }>({})
+
+  const { data: projects } = useProjectsQuery()
+  const { data: organizations } = useOrganizationsQuery()
+  const {
+    data,
+    error,
+    isLoading,
+    isError,
+    isSuccess,
+    hasNextPage,
+    isFetchingNextPage,
+    fetchNextPage,
+  } = useNotificationsV2Query({
+    status: activeTab === 'archived' ? 'archived' : selectedFilter === 'unread' ? 'new' : undefined,
+    priority:
+      selectedFilter === 'critical'
+        ? 'Critical'
+        : selectedFilter === 'warning'
+        ? 'Warning'
+        : undefined,
+  })
+  const { data: summary } = useNotificationsSummaryQuery()
+  const { mutate: updateNotifications } = useNotificationsV2UpdateMutation()
+
+  const notifications = useMemo(() => data?.pages.flatMap((page) => page) ?? [], [data?.pages])
+  const hasNewNotifications = summary?.unread_count ?? 0 > 0
+  const hasWarning = summary?.has_warning
+  const hasCritical = summary?.has_critical
+
+  const onSelectFilter = (value: 'all' | 'unread' | 'warning' | 'critical') => {
+    setSelectedFilter(value)
+    setOpenFilters(false)
+  }
+
+  const markNotificationsRead = () => {
+    const ids = notifications.filter((n) => n.status === 'new').map((n) => n.id)
+    updateNotifications({ ids, status: 'seen' })
+  }
+
+  return (
+    <Popover_Shadcn_
+      modal={false}
+      open={open}
+      onOpenChange={(open) => {
+        setOpen(open)
+        if (!open) markNotificationsRead()
```

Comment:
For making notifications as read, what do you think of having each notification manage it's own read state? So instead of making them all read when you open the popover you do something like this on each row:
```
import { useInView } from 'react-intersection-observer'

const { ref, inView } = useInView()

useEffect(() => {
  if (inView) {
    markRead()
  }
}, [inView])

<div ref={ref}>
```

That way you can't accidentally mark notifications as read if they're out of view.

Reply:
great point this - and a better behaviour, i'll look into it 🙏